README.md
**Resumable HTTP download request** - pause, resume, cancel, realtime progress and error handling

## Features

- **Take control** - pause, resume and cancel
- **Listen to updates** - realtime progress and failure handling
- **Pure Dart** - only `http` dependency 
- **Easy to use** - singletone and stream

## Getting started

Include latest version from [pub.dev](https://pub.dev/packages/download_task) to `pubspec.yaml` and simply run
```dart
await DownloadTask.download(url, destination);
```

## Usage

```dart
// initialize download request
final task = await DownloadTask.download(url, File("image.webp"));

// listen to state changes
task.events.listen((event) { ... }

// control task
task.pause();
task.resume();
task.cancel();
```
Example full source code available at [example's directory](example/download_task_example.dart)


## Additional information

This package is primarly used in [isolated_download_manager](https://pub.dev/packages/isolated_download_manager)


CHANGELOG.md
## 1.0.0

- Initial version

## 1.0.1

- Source code readability improvements
- Clean up an example

## 1.0.2

- Added missing headers in download method by @Mithos5r

## 1.0.3

- Update dependencies for the Dart 3 support

## 1.0.4

- Issue with fast switching between paused/downloading modes resolved

test/download_task_test.dart
// @Timeout(Duration(seconds: 300))
import 'dart:async';

import 'package:download_task/download_task.dart';
import 'package:http/testing.dart';
import 'package:test/test.dart';
import 'package:http/http.dart' as http;
import 'dart:io' show File, Directory;

// Code Coverage
// flutter test --coverage
// genhtml coverage/lcov.info -o coverage/html
//
// Format Code
// dart format lib
//
// Package Analysis
// https://github.com/dart-lang/pana
// pana .

final List<Uri> links = [
  Uri.parse("https://download.task/image.png"),
  Uri.parse("https://download.task/video.mp4"),
  Uri.parse("https://download.task/notFound.zip"),
];
String rootDirectory = "${Directory.current.path}/test";
final String downloadDirectory = "$rootDirectory/temp";

void main() {
  final client = MockClient.streaming((request, bodyStream) {
    if (request.url == links[0]) {
      return streamFile("image.png");
    } else if (request.url == links[1]) {
      // return streamRandomBytes(50 * 1024 * 1024);
      return longStream(seconds: 10);
      // return streamFile("video.mp4");
    } else if (request.url == links[2]) {
      // final response = http.StreamedResponse(Stream.fromIterable([]), 404);
      // return Future.value(response);
      throw UnimplementedError();
    }
    throw UnimplementedError();
  });

  group("Valid simple", () {
    late DownloadTask task;
    final file = File("$downloadDirectory/image.png");

    setUp(() async {
      task = await DownloadTask.download(links[0], file: file, client: client);
    });

    tearDown(() async {
      deleteFile(file);
    });

    test("Downloaded", () async {
      final last = await task.events.last;
      expect(last.state, equals(TaskState.success));

      expect(file.existsSync(), equals(true));
    });
  });
  group("Valid with controller", () {
    late DownloadTask task;
    final file = File("$downloadDirectory/video.mp4");

    setUp(() async {
      task = await DownloadTask.download(links[1], file: file, client: client);
    });

    tearDown(() async {
      deleteFile(file);
    });

    test("Pause, Resume, Cancel", () async {
      Future.delayed(const Duration(milliseconds: 150)).then((_) => task.pause());
      Future.delayed(const Duration(milliseconds: 300)).then((_) => task.resume());
      Future.delayed(const Duration(milliseconds: 450)).then((_) => task.pause());
      Future.delayed(const Duration(milliseconds: 600)).then((_) => task.resume());
      Future.delayed(const Duration(milliseconds: 750)).then((_) => task.cancel());

      final events = await task.eventStatesSequence();

      expect(events, equals([
        TaskState.downloading,
        TaskState.paused,
        TaskState.downloading,
        TaskState.paused,
        TaskState.downloading,
        TaskState.canceled,
      ]));
    });
  });
  
  group("Advanced", () {
    late DownloadTask task;
    final file = File("$downloadDirectory/video.mp4");

    setUp(() async {
      // file which already exists
      await file.create();
      final bytes = List.generate(1024, (index) => 100);
      await file.writeAsBytes(bytes);

      task = await DownloadTask.download(links[1], file: file, client: client);
    });

    tearDown(() async {
      deleteFile(file);
    });

    test("Continue downloading when file exists ", () async {
      expect(task.event?.bytesReceived, equals(-1));
      await Future.delayed(const Duration(milliseconds: 900));
      // file exists with 1024 bytes and 3 bytes per second were downloaded
      expect(task.event?.bytesReceived, equals(1027));
    });
  });

  group("Bad link", () {
    late DownloadTask task;
    final file = File("$downloadDirectory/file.zip");

    setUp(() async {
      task = await DownloadTask.download(links[2], file: file, client: client, deleteOnError: true);
    });

    tearDown(() async {
      deleteFile(file);
    });

    test("Should fail", () async {
      await Future.delayed(const Duration(milliseconds: 300));
      final events = await task.events.toList();
      // print(events);
      expect(events.last.state, equals(TaskState.error));      
      expect(file.existsSync(), equals(false));
    });
  });
}

Future<http.StreamedResponse> streamFile(String filename){
  final file = File("$rootDirectory/files/$filename");
  final stream = file.openRead();
  final response = http.StreamedResponse(stream, 200);
  return Future.value(response);
}

// Uint8List(1024)
/*Future<http.StreamedResponse> streamRandomBytes(int length) {
  final list = List.generate(length, (index) => 100);
  final stream = Stream.fromIterable([list]);
  final response = http.StreamedResponse(stream, 200);
  return Future.value(response);
}*/

Future<http.StreamedResponse> longStream({ required int seconds }) {
  final controller = StreamController<List<int>>();
  for (int i = 0; i <= seconds; i++) {
    Future.delayed(Duration(seconds: i))
      .then((_) => controller.add([100, 100, 100]));
  }
  final response = http.StreamedResponse(controller.stream, 200);
  return Future.value(response);
}


Future<void> deleteFile(File file) async {
  if (await file.exists()) {
    await file.delete();
  }
}

extension DeduplicateStates on DownloadTask {
  // convert stream of events to states list and remove duplicates (only when appears in a row)
  Future<List<TaskState>> eventStatesSequence() async {
   final eventsList = await events.map((e) => e.state).toList();   
    List<TaskState> uniqueSequence = []; 
    for (int i = 0; i < eventsList.length; i++) {
      final event = eventsList[i];
      if (i == 0) {
        uniqueSequence.add(event);
        continue;
      }
      if (event != eventsList[i-1]) {
        uniqueSequence.add(event);
      }
    }
    return uniqueSequence;
  }
}

extension SkipProgressEvents on Stream<TaskEvent> {
  Stream<TaskEvent> skipDownloading() {
    return skipWhile((e) => e.state == TaskState.downloading);
  }
}

test/resumable_download_test.dart
import 'package:resumable_download/resumable_download.dart';
import 'package:test/test.dart';

void main() {
  /*group('A group of tests', () {
    final awesome = Awesome();

    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {
      expect(awesome.isAwesome, isTrue);
    });
  });*/
}

lib/src/task.dart
import 'dart:async' show StreamController, StreamSubscription;
import 'dart:io' show File, FileMode;
import 'package:http/http.dart' as http;

/// Downloading state
enum TaskState { downloading, paused, success, canceled, error }

/// Event representing current progress or error and the current state
class TaskEvent {
  const TaskEvent(
      {required this.state, this.bytesReceived, this.totalBytes, this.error});

  final TaskState state;
  final int? bytesReceived;
  final int? totalBytes;
  final Object? error;

  @override
  String toString() => "TaskEvent ($state)";
}

/// Main class, used as a singletone for initialising the downloads
class DownloadTask {
  DownloadTask._({
    required this.url,
    required this.file,
    required this.headers,
    required this.client,
    required this.deleteOnCancel,
    required this.deleteOnError,
    this.size,
    this.safeRange = false,
  });

  final Uri url;
  final File file;
  final Map<String, String> headers;
  final http.Client client;
  final bool deleteOnCancel;
  final bool deleteOnError;
  final int? size;
  final bool safeRange;

  /// Events stream, used to listen for downloading state changes
  Stream<TaskEvent> get events => _events.stream;

  /// Latest event
  TaskEvent? get event => _event;

  /// Static method to fire file downloading returns future of [DownloadTask] which may be used to control the request
  ///
  /// * [headers] are custom HTTP headers for client, may be used for request authentication
  /// * if [client] is `null` the default one will be used
  /// * [file] is download path, file will be created while downloading
  /// * [deleteOnCancel] specify if file should be deleted after download is cancelled
  /// * [deleteOnError] specify if file should be deleted when error is raised
  /// * [size] used to specify bytes end for range header
  /// * [safeRange] used to skip range header if bytes end not found
  ///
  static Future<DownloadTask> download(
    Uri url, {
    Map<String, String> headers = const {},
    http.Client? client,
    required File file,
    bool deleteOnCancel = true,
    bool deleteOnError = false,
    int? size,
    bool safeRange = false,
  }) async {
    final task = DownloadTask._(
        url: url,
        headers: headers,
        client: client ?? http.Client(),
        file: file,
        deleteOnCancel: deleteOnCancel,
        deleteOnError: deleteOnError,
        size: size,
        safeRange: safeRange);
    await task.resume();
    return task;
  }

  /// Pause file downloading, file will be stored on defined location
  /// downloading may be continued from the paused point if file exists
  Future<bool> pause() async {
    if (_doneOrCancelled || !_downloading) return false;
    _addEvent(TaskEvent(
        state: TaskState.paused,
        bytesReceived: _bytesReceived,
        totalBytes: _totalBytes));
    await _subscription?.cancel();
    _subscription = null;
    return true;
  }

  /// Resume file downloading, if file exists downloading will continue from file size
  /// will return `false` if downloading is in progress, finished or cancelled
  Future<bool> resume() async {
    if (_doneOrCancelled || _downloading) return false;
    _addEvent(TaskEvent(
        state: TaskState.downloading,
        bytesReceived: _bytesReceived,
        totalBytes: _totalBytes));
    _download().then((value) => _subscription = value);
    // _subscription = await  _download();
    return true;
  }

  /// Cancel the downloading, if [deleteOnCancel] is `true` then file will be deleted
  /// will return `false` if downloading was already finished or cancelled
  Future<bool> cancel() async {
    if (_doneOrCancelled) return false;
    _addEvent(TaskEvent(
        state: TaskState.canceled,
        bytesReceived: _bytesReceived,
        totalBytes: _totalBytes));
    await _subscription?.cancel();
    _subscription = null;
    _dispose(TaskState.canceled);
    return true;
  }

  // Events stream
  StreamSubscription? _subscription;
  final StreamController<TaskEvent> _events = StreamController<TaskEvent>();
  TaskEvent? _event;

  int _bytesReceived = -1;
  int _totalBytes = -1;

  // Internal shortcuts
  bool get _cancelled => event?.state == TaskState.canceled;
  bool get _downloading => event?.state == TaskState.downloading;
  bool get _done => event?.state == TaskState.success;
  bool get _doneOrCancelled => _done || _cancelled;

  /// Add new event to stream
  void _addEvent(TaskEvent event) {
    _event = event;
    if (!_events.isClosed) {
      _events.add(event);
    }
  }

  /// Clean up
  Future<void> _dispose(TaskState state) async {
    if (state == TaskState.canceled) {
      if (deleteOnCancel) {
        await file.delete();
      }
      _events.close();
    } else if (state == TaskState.error) {
      if (deleteOnError) {
        await file.delete();
      }
      _events.close();
    } else if (state == TaskState.success) {
      _events.close();
    }
  }

  /// Download function
  /// returns future of [StreamSubscription] which used to receive updates internally
  /// returns `null` on error
  Future<StreamSubscription?> _download() async {
    late final StreamSubscription subscription;

    Future<void> onError(Object error) async {
      _addEvent(TaskEvent(state: TaskState.error, error: error));
      _dispose(TaskState.error);
    }

    try {
      // calculate starting point
      int from = _bytesReceived;
      if (from == -1) {
        if (await file.exists()) {
          // use existed file bytes count
          from = await file.length();
        } else {
          from = 0;
          await file.create(recursive: false);
        }
      }
      final sink = await file.open(mode: FileMode.writeOnlyAppend);

      final request = http.Request("GET", url);

      request.headers.addAll(headers);

      // range header
      if (size != null) {
        request.headers["Range"] = "bytes=$from-$size";
      } else {
        if (!safeRange) {
          request.headers["Range"] = "bytes=$from-";
        }
      }

      final response = await client.send(request);

      // length total
      int totalBytes = -1;
      final length = response.contentLength;
      if (length != null) {
        totalBytes = from + length;
      } else {
        // Content-Lenght header is missing, a try to get size from Content-Range header
        final range = response.headers["content-range"];
        if (range != null) {
          final index = range.indexOf("/");
          if (index != -1) {
            final total = int.tryParse(range.substring(index + 1));
            if (total != null) totalBytes = total;
          }
        }
      }

      // process
      subscription = response.stream.listen((data) async {
        if (subscription != _subscription) return;
        subscription.pause();
        await sink.writeFrom(data);
        final bytesReceived = from + data.length;
        from = bytesReceived;
        _bytesReceived = bytesReceived;
        _totalBytes = totalBytes;

        if (_doneOrCancelled ||
            !_downloading ||
            subscription != _subscription) {
          subscription.cancel();
        } else {
          _addEvent(TaskEvent(
              state: TaskState.downloading,
              bytesReceived: bytesReceived,
              totalBytes: totalBytes));
          subscription.resume();
        }
      }, onDone: () async {
        _addEvent(const TaskEvent(state: TaskState.success));
        await sink.close();
        // client.close();
        _dispose(TaskState.success);
      }, onError: onError);

      return subscription;
    } catch (error) {
      await onError(error);
      return null;
    }
  }
}


lib/src/config.dart
/*part of 'package:download_task/src/download.dart';

class DownloadConfig {
  const DownloadConfig._({
    required this.url,
    required this.file,
    required this.headers,
    required this.client,
    this.onProgress,
    this.onDone,
    this.onError,
    this.onCancel,
    this.onPause,
    this.onResume,
    required this.deleteOnCancel,
    required this.deleteOnError,
  });

  final Uri url;
  final File file;
  final Map<String, String> headers;
  final http.Client client;

  final void Function(int bytesReceived, int totalBytes)? onProgress;
  final void Function()? onDone;
  final void Function(Object)? onError;
  final void Function()? onCancel;
  final void Function(int bytesReceived, int totalBytes)? onPause;
  final void Function()? onResume;

  final bool deleteOnCancel;
  final bool deleteOnError;
}*/

lib/src/controller.dart
part of 'package:resumable_download/src/resumable_download_base.dart';

class DownloadController {
  DownloadController._({
    required StreamSubscription? subscription, 
    required DownloadConfig args,
  }) : _subscription = subscription, _args = args {

    final onProgress = args.onProgress;
    args.onProgress = (bytesReceived, totalBytes) {
      _bytesReceived = bytesReceived;
      _totalBytes = totalBytes;
      onProgress?.call(bytesReceived, totalBytes);
    };

    final onDone = args.onDone;
    args.onDone = () {
      _done = true;
      onDone?.call();
    };
  }

  final DownloadConfig _args;
  StreamSubscription? _subscription;
  int _bytesReceived = -1;
  int _totalBytes = -1;

  bool _cancelled = false;
  bool get isCancelled => _cancelled;

  bool _downloading = true;
  bool get isDownloading => _cancelled;

  bool _done = false;
  bool get isDone => _done;

  bool get doneOrCancelled => isDone || _cancelled;

  Future<bool> pause() async {
    if (doneOrCancelled || !_downloading) return false;
    await _subscription?.cancel();
    _downloading = false;
    _args.onPause?.call(_bytesReceived, _totalBytes);
    return true;
  }

  Future<bool> resume() async {
    if (doneOrCancelled || _downloading) return false;
    _subscription = await _download(args: _args, from: _bytesReceived);
    _downloading = true;
    _args.onResume?.call();
    return true;
  }

  Future<bool> cancel() async {
    if (doneOrCancelled) return false;
    // _cancelled = true;
    await _subscription?.cancel();
    _cancelled = true;
    if (_args.deleteOnCancel) {
      await _args.file.delete();
    }
    _args.onCancel?.call();
    return true;
  }
}

example/resumable_download_example.dart
import 'package:resumable_download/resumable_download.dart' as resumable;
import 'dart:io' show File;

void main() async {
  final url = Uri.parse(
    // "https://not.found/html.txt",
    "https://golang.org/dl/go1.17.3.src.tar.gz",
  );

  double previousProgress = 0.0;
  final request = await resumable.download(
    url: url,
    // headers: { }
    file: File("/Users/starkdmi/Downloads/test/Go.gz"),
    onProgress: (bytesReceived, totalBytes) {
      if (totalBytes == -1) return;
      final progress = (bytesReceived / totalBytes * 100).floorToDouble();
      if (progress != previousProgress && progress % 10 == 0) {
        print("progress $progress%");
        previousProgress = progress;
      }
    },
    onDone: () => print("downloaded"),
    onError: (error) => print("error: $error"),
    onCancel: () => print("onCancel - cancelled"),
    onPause: (bytesReceived, totalBytes) => print("onPause - paused"),
    // deleteOnCancel: true,
    // deleteOnError: true,
  );

  await Future.delayed(const Duration(milliseconds: 500));
  request.pause().then((status) => print(status ? "paused" : "can't pause"));
  // await Future.delayed(const Duration(milliseconds: 500));
  // request.pause().then((status) => print(status ? "paused" : "can't pause"));
  await Future.delayed(const Duration(milliseconds: 500));
  request.resume().then((status) => print(status ? "resumed" : "can't resume"));
  await Future.delayed(const Duration(milliseconds: 1500));
  request.cancel().then((status) => print(status ? "cancelled" : "can't cancell"));
}


example/download_task_example.dart
import 'package:download_task/download_task.dart';
import 'dart:io' show File, Directory;

void main() async {
  // specify url and destionation
  final url = Uri.parse("https://golang.org/dl/go1.19.1.src.tar.gz");
  final file = File("${Directory.current.path}/example/Go.tar.gz");

  // initialize download request
  final task = await DownloadTask.download(url, file: file);

  // listen to state changes
  double previousProgress = 0.0;
  task.events.listen((event) {
    switch (event.state) {
      case TaskState.downloading:
        final bytesReceived = event.bytesReceived!;
        final totalBytes = event.totalBytes!;
        if (totalBytes == -1) return;

        final progress = (bytesReceived / totalBytes * 100).floorToDouble();
        if (progress != previousProgress && progress % 10 == 0) {
          print("progress $progress%");
          previousProgress = progress;
        }
        break;
      case TaskState.paused:
        print("paused");
        break;
      case TaskState.success:
        print("downloaded");
        break;
      case TaskState.canceled:
        print("canceled");
        break;
      case TaskState.error:
        print("error: ${event.error!}");
        break;
    }
  });

  // control the process
  await Future.delayed(const Duration(milliseconds: 500));
  task.pause();
  await Future.delayed(const Duration(milliseconds: 500));
  task.resume();
  await Future.delayed(const Duration(milliseconds: 1500));
  task.cancel();
}
